okhttp/src/main/java/io/grpc/okhttp/OkHttpServerTransport.java

```diff
@@ -638,6 +644,11 @@ public void headers(boolean outFinished,
         newStream = streamId > lastStreamId;
         if (newStream) {
           lastStreamId = streamId;
+          if (config.maxConcurrentStreams <= streams.size()) {
+            streamError(streamId, ErrorCode.REFUSED_STREAM,
```

Comment:
PROTOCOL_ERROR would handle implementation bugs better, because the streams wouldn't be retried. But there is a handshake race for new connections where the client may initially exceed the stream limit. REFUSED_STREAM is most helpful if the stream limit is changing over the life of the connection. To make this "really nice" we could set a flag in `ackSettings()` and do `receivedSettingsAck ? ErrorCode.PROTOCOL_ERROR : ErrorCode.REFUSED_STREAM`, because there's no valid reason for the client to exceed the stream limit after acking the settings.

But what you have is fine, and that'd be more annoying for you to test.